README.md
# leaflet-overpass-lens
On a Leaflet Map, query an area around the mouse pointer for map items via Overpass API (an OpenStreetMap database server)

-> [DEMO](https://xover.mud.at/~skunk/leaflet-overpass-lens/)

## Screenshot
![Screenshot of the demo application, showing the configuration on the left side. On the right side a map with buildings and roads intersecting the selected lens.](screenshot.png)

See also this short screencast: https://youtu.be/JTzVa9E0Yo8

## Usage
### Use demo application
```sh
git clone https://github.com/plepe/leaflet-overpass-lens
cd leaflet-overpass-lens
npm install
npm start
```

Visit http://localhost:8080

For developing, additionally run `npm run watch`. This will include debug information in the compiled package (you can use the browser developer tools for debugging) and will re-compile whenever the code changes.

### Plain HTML/JS
Run on a shell:
```sh
npm install leaflet-overpass-lens leaflet
```

Create a HTML file `index.html`:
```
<html>
<head>
  <link rel="stylesheet" href="node_modules/leaflet/dist/leaflet.css"/>
  <link rel="stylesheet" href="node_modules/leaflet-overpass-lens/dist/leaflet-overpass-lens.css"/>
  <script src="node_modules/leaflet/dist/leaflet.js"></script>
  <script src="node_modules/leaflet-overpass-lens/dist/leaflet-overpass-lens.js"></script>
</head>
<body>
  <div id="map" style="position: absolute; left: 0; right: 0; top: 0; bottom: 0;"></div>
  <script>
// use the code from the "Code" section below
  </script>
</body>
</html>
```

You might also need a web server, e.g. Apache or [http-server](https://www.npmjs.com/package/http-server).

### With browserify (or other packaging tools)
Create/update your package.json:
```json
{
  "scripts": {
    "build": "browserify -g browserify-css app.js -o dist/app.js"
  }
}
```

Run on a shell:
```sh
npm install --save leaflet-overpass-lens leaflet overpass-frontend browserify
```

Create a file app.js with the following code:
```js
const OverpassFrontend = require('overpass-frontend')
require('leaflet')
require('leaflet-overpass-lens')

// use the code from the "Code" section below
```

Create a HTML file `index.html`:
```
<html>
  <head>
    <link rel="stylesheet" href="node_modules/leaflet/dist/leaflet.css"/>
    <link rel="stylesheet" href="node_modules/leaflet-overpass-lens/dist/leaflet-overpass-lens.css"/>
  </head>
  <body>
    <div id="map" style="position: absolute; left: 0; right: 0; top: 0; bottom: 0;"></div>
    <script src="dist/app.js"></script>
  </body>
</html>
```

Build your application:
```sh
npm run build
```

You might also need a web server, e.g. Apache or [http-server](https://www.npmjs.com/package/http-server).

### Code
```js
const overpassURL = '//overpass-api.de/api/interpreter'
// const overpassURL = 'map.osm' // Download a .osm file to this directory and use this instead
const overpassFrontend = new OverpassFrontend(overpassURL)

const map = L.map('map').setView([51.505, -0.09], 15)

// Map Background
L.tileLayer('//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', { attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>' }
).addTo(map)

// you can omit these values; shown are the default values.
const options = {
  position: 'topleft', // position of the control
  continuous: false, // if true, lens will follow mouse and update continuously
  icon: '<span>🔍</span>', // which icon to use on the control
  radius: 100, // radius around the mouse pointer
  radiusUnits: 'meters', // units
  bufferStyle: { weight: 3, color: '#007fff', fill: false } // the visible buffer around the mouse position
}

// Configure the overlay. For more options, please check https://github.com/plepe/overpass-layer#readme
const layerOptions = {
  overpassFrontend: overpassFrontend,
  minZoom: 15,
  feature: { markerSymbol: '' }, // hide markers, image won't be found
  query: 'nwr', // load everything in that area
}

L.overpassLens(options, layerOptions).addTo(map)
```


CHANGELOG.md
Version 0.1.0 - 2021-12-31
==========================
* First public version


src/index.js
/* global L:false */
const OverpassLayer = require('overpass-layer')
const turf = {
  buffer: require('@turf/buffer').default
}

const defaultOptions = {
  position: 'topleft',
  continuous: false,
  icon: '<span>🔍</span>',
  radius: 100,
  radiusUnits: 'meters',
  bufferStyle: { weight: 3, color: '#007fff', fill: false }
}

L.OverpassLens = L.Control.extend({
  initialize: function (options = {}, layerOptions = {}) {
    this.layerOptions = layerOptions

    this.options = defaultOptions
    L.setOptions(this, options)
    L.Control.prototype.initialize.call(this, options)
  },

  onAdd: function (map) {
    this.map = map

    const container = L.DomUtil.create('div', 'leaflet-bar leaflet-control-overpass-lens')
    const containerLink = document.createElement('a')
    containerLink.href = '#'
    containerLink.innerHTML = this.options.icon
    containerLink.title = 'Query map'
    container.appendChild(containerLink)
    this.container = container

    this.map.on('mousemove', (e) => {
      if (e && this.options.continuous) {
        this.position = e.latlng

        if (this.isShown) {
          const geom = this.geometry(e.latlng)
          this.layer.setBounds(geom)

          this.buffer.clearLayers()
          this.buffer.addData(geom)
        }
      }
    })

    containerLink.onclick = () => {
      if (this.isShown) {
        return this.hide()
      }

      setTimeout(() => {
        if (this.options.continuous) {
          this.show()
        } else {
          const origCursor = this.map._container.style.cursor
          this.map._container.style.cursor = 'help'
          container.classList.add('enabled')

          this.map.once('click', (e) => {
            this.map._container.style.cursor = origCursor

            this.position = e.latlng
            this.show()
          })
        }
      }, 0)

      return false
    }

    return container
  },

  geometry (position) {
    return turf.buffer({
      type: 'Feature',
      geometry: {
        type: 'Point',
        coordinates: [position.lng, position.lat]
      }
    }, this.options.radius || 200, { units: this.options.radiusUnits || 'meters' })
  },

  show () {
    this.container.classList.remove('enabled')
    this.container.classList.add('active')

    const geom = this.geometry(this.position)

    if (!this.layer) {
      this.layerOptions.bounds = geom
      this.layer = new OverpassLayer(this.layerOptions)

      this.buffer = L.geoJSON(geom, this.options.bufferStyle)
    } else {
      this.layer.setBounds(geom)

      this.buffer.clearLayers()
      this.buffer.addData(geom)
    }

    this.buffer.addTo(this.map)
    this.layer.addTo(this.map)
    this.isShown = true
  },

  updateOptions (options, layerOptions) {
    this.options = defaultOptions
    L.setOptions(this, options)
    this.layerOptions = layerOptions

    if (this.isShown) {
      this.hide()
      this.layer = null
      this.show()
    } else {
      this.layer = null
    }
  },

  hide () {
    this.container.classList.remove('enabled')
    this.container.classList.remove('active')

    this.layer.remove()
    this.buffer.remove()
    this.isShown = false
  }
})

L.overpassLens = function (overpass, options) {
  return new L.OverpassLens(overpass, options)
}


demo.js
/* global L:false, OverpassFrontend:false */
const overpassURL = '//overpass-api.de/api/interpreter'
// var overpassURL = 'map.osm' // Download a .osm file to this directory and use this instead
const overpassFrontend = new OverpassFrontend(overpassURL)

const map = L.map('map').setView([51.505, -0.09], 15)

const osmMapnik = L.tileLayer('//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
  {
    maxZoom: 19,
    attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>'
  }
)
osmMapnik.addTo(map)

let lens

function start () {
  const options = {
    continuous: false,
    radius: 100,
    radiusUnits: 'meters'
  }

  const layerOptions = {
    overpassFrontend: overpassFrontend,
    query: 'nwr',
    minZoom: 15,
    feature: {
      body: '',
      feature: {
        style: {}
      },
      markerSymbol: ''
    }
  }

  const form = document.getElementById('options')
  Array.from(form.elements).forEach(el => {
    const path = el.name.split('.')

    if (path[0] === 'options') {
      options[path[1]] = inputValue(el)
    } else if (path[0] === 'layerOptions') {
      if (path.length === 4) {
        if (!layerOptions[path[1]][path[2]]) {
          layerOptions[path[1]][path[2]] = {}
        }
        if (!layerOptions[path[1]][path[2]][path[3]]) {
          layerOptions[path[1]][path[2]][path[3]] = {}
        }

        layerOptions[path[1]][path[2]][path[3]] = inputValue(el)
      } else if (path.length === 3) {
        if (!layerOptions[path[1]][path[2]]) {
          layerOptions[path[1]][path[2]] = {}
        }

        layerOptions[path[1]][path[2]] = inputValue(el)
      } else {
        layerOptions[path[1]] = inputValue(el)
      }
    }
  })

  if (lens) {
    lens.updateOptions(options, layerOptions)
  } else {
    lens = L.overpassLens(options, layerOptions)
    lens.addTo(map)
  }
}

function change () {
  start()
}

function init () {
  const form = document.getElementById('options')
  Array.from(form.elements).forEach(el => {
    el.onchange = () => change()
  })

  start()
}

function inputValue (element) {
  if (element.type === 'checkbox') {
    return !!element.checked
  }

  return element.value
}

init()
